### PR TITLE
Custom equality comparer in OnChange resolves #156

### LINKED ISSUE #156: Allow for custom equality comparer in Agenda Filter's OnKeyChangePredicate
```csharp
// typeof(key) is TKey
Filter()
    .OnChange(() => key, () => CustomTKeyEqualityComparer.Instance);
```

### PR DESCRIPTION
**Pull request by linanqiu**

### CODE REVIEW

**Comment by linanqiu** [comment]
This could also be an extension method on `IFilterExpression` but I don't see any pattern of building "secondary" extensions on existing expressions within this project itself

**Comment by linanqiu** [comment]
Feel free to reject this if you don't think this will be useful for many users

**Comment by snikolayev** [comment]
I'm still not convinced this is a good engine level capability. I'll review the changes and think about this, and maybe have more questions.

File: src/NRules/NRules.Fluent/Dsl/IFilterExpression.cs [hunk 1]
```diff
@@ -16,6 +17,16 @@ public interface IFilterExpression
...
 
+        /// <summary>
+        /// Configures the engine to filter rule's matches, so that updates are only triggered if given key changed.
+        /// Comparison is done using the custom equality comparer given
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="keySelector">Key selector expression.</param>
+        /// <param name="comparer">Equality comparer for key</param>
+        /// <returns></returns>
+        IFilterExpression OnChange<T>(Expression<Func<T>> keySelector, Expression<Func<IEqualityComparer<T>>> comparer) where T : class;
```
Comment (snikolayev) [inline src/NRules/NRules.Fluent/Dsl/IFilterExpression.cs:28]: Why should the comparer be an expression? It seems sufficient to just pass the actual comparer instance. Or am I missing something?

File: src/NRules/NRules.Fluent/Expressions/FilterExpression.cs [hunk 2]
```diff
@@ -34,7 +44,45 @@ public IFilterExpression Where(params Expression<Func<bool>>[] predicates)
                 var expression = predicate.DslExpression(filters.Declarations);
                 filters.Filter(FilterType.Predicate, expression);
             }
+
             return this;
         }
+
+        private static Expression<Func<object>> InternalizeComparer<T>(Expression<Func<T>> keySelector, Expression<Func<IEqualityComparer<T>>> comparer) where T : class
```
Comment (snikolayev) [inline src/NRules/NRules.Fluent/Expressions/FilterExpression.cs:51]: This seems pretty hacky. I don't think the equality wrapper should be part of the engine, especially given the restriction on the type argument to be a reference type

Comment (linanqiu) [reply src/NRules/NRules.Fluent/Expressions/FilterExpression.cs:51]: Agreed. I'll implement this as an extension in my application then.

**Comment by linanqiu** [comment]
A

**Pull request closed without merging**